samples/bluetooth/hci_uart/src/main.c

```diff
@@ -49,78 +55,136 @@ static K_FIFO_DEFINE(tx_queue);
+	return rx;
 }
 
-static size_t h4_discard(struct device *uart, size_t len)
+static bool valid_type(uint8_t type)
 {
-	uint8_t buf[H4_DISCARD_LEN];
-
-	return uart_fifo_read(uart, buf, MIN(len, sizeof(buf)));
+	return (type == H4_CMD) | (type == H4_ACL);
 }
 
-static void h4_cmd_recv(struct net_buf *buf, int *remaining)
+/* Function assumes that type is validated and only CMD or ACL will be used. */
+static uint32_t get_len(union bt_hci_hdr *hdr, uint8_t type)
 {
-	struct bt_hci_cmd_hdr hdr;
-
-	/* We can ignore the return value since we pass len == min */
-	h4_read(hci_uart_dev, (void *)&hdr, sizeof(hdr), sizeof(hdr));
-
-	*remaining = hdr.param_len;
-
-	net_buf_add_mem(buf, &hdr, sizeof(hdr));
+	return (type == BT_BUF_CMD) ?
+		hdr->cmd.param_len : sys_le16_to_cpu(hdr->acl.len);
+}
 
-	LOG_DBG("len %u", hdr.param_len);
+/* Function assumes that type is validated and only CMD or ACL will be used. */
+static int hdr_len(uint8_t type)
+{
+	return (type == H4_CMD) ?
+		sizeof(struct bt_hci_cmd_hdr) : sizeof(struct bt_hci_acl_hdr);
 }
 
-static void h4_acl_recv(struct net_buf *buf, int *remaining)
+static void rx_isr(void)
 {
-	struct bt_hci_acl_hdr hdr;
+	static struct net_buf *buf;
+	static int remaining;
+	static uint8_t state;
+	static uint8_t type;
+	static union bt_hci_hdr hdr;
+	int read;
+
+	do {
+		switch (state) {
+		case ST_IDLE:
+			/* Get packet type */
+			read = h4_read(hci_uart_dev, &type, sizeof(type));
+			if (read) {
+				if (valid_type(type)) {
+					/* Get expected header size and switch
+					 * to receiving header.
+					 */
+					remaining = hdr_len(type);
+					state = ST_HDR;
+				} else {
+					LOG_WRN("Unknown header %d", state);
+				}
+			}
+			break;
+		case ST_HDR:
+			read = h4_read(hci_uart_dev,
+				       &hdr.raw[hdr_len(type) - remaining],
+				       remaining);
+			remaining -= read;
+			if (remaining == 0) {
+				/* Header received. Get payload length and
+				 * allocate buffer. If allocation fails, discard
+				 * packet.
+				 */
+				remaining = get_len(&hdr, type);
+				buf = bt_buf_get_tx(BT_BUF_H4, K_NO_WAIT,
+						&type, sizeof(type));
+				if (buf == NULL) {
+					return;
+				}
+
+				net_buf_add_mem(buf, hdr.raw, hdr_len(type));
+				if (remaining > net_buf_tailroom(buf)) {
+					LOG_ERR("Not enough space in buffer");
+					net_buf_unref(buf);
+					state = ST_DISCARD;
+				} else {
+					state = ST_PAYLOAD;
+				}
+
+			}
+			break;
+		case ST_PAYLOAD:
+			read = h4_read(hci_uart_dev, net_buf_tail(buf),
+					remaining);
+			buf->len += read;
+			remaining -= read;
+			if (remaining == 0) {
+				/* Packet received */
+				LOG_DBG("putting RX packet in queue.");
+				net_buf_put(&tx_queue, buf);
+				state = ST_IDLE;
+			}
+			break;
+		case ST_DISCARD:
+		{
+			uint8_t discard[H4_DISCARD_LEN];
 
-	/* We can ignore the return value since we pass len == min */
-	h4_read(hci_uart_dev, (void *)&hdr, sizeof(hdr), sizeof(hdr));
+			read = h4_read(hci_uart_dev, discard, sizeof(discard));
```

Comment:
Shouldn't it rather read `MIN(remaining, sizeof(discard))` as the original code did?

Reply:
done